diff --unified --recursive --text a/module/spl/spl-vnode.c b/module/spl/spl-vnode.c
--- a/module/spl/spl-vnode.c	2019-02-22 12:59:44.000000000 -0500
+++ b/module/spl/spl-vnode.c	2019-05-15 18:48:43.392458053 -0400
@@ -608,7 +608,7 @@
 	 * size to ensure strncpy_from_user() does not fail with -EFAULT.
 	 */
 	saved_fs = get_fs();
-	set_fs(get_ds());
+	set_fs(KERNEL_DS);
 
 	rc = user_path_dir(filename, &path);
 	if (rc)
